### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by shm04**
### In this pull request we made the following changes:

   - Make sure that the N+1 problem is solved when fetching all posts and their comments for a user

- User index page:
  - I can see the username of all other users.
  - I can see the profile picture for each user.
  - I can see the number of posts each user has written.
  - When I click on a user, I am redirected to that user's show page.

- Users index (before)

image

- Users index (after)

image

- Users show (before)

image

- Users show (after)

image

- Posts index (before)

image

-Posts index (after)

image

- Posts show (index)

image

- Posts show (after)

image

- user show page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see the user's bio.
  - I can see the user's first 3 posts.
  - I can see a button that lets me view all of a user's posts.
  - When I click a user's post, it redirects me to that post's show page.
  - When I click to see all posts, it redirects me to the user's post's index page.
 
- User post index page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see a post's title.
  - I can see some of the post's body.
  - I can see the first comments on a post.
  - I can see how many comments a post has.
  - I can see how many likes a post has.
  - I can see a section for pagination if there are more posts than fit on the view.
  - When I click on a post, it redirects me to that post's show page.

- Post show page:
  - I can see the post's title.
  - I can see who wrote the post.
  - I can see how many comments it has.
  - I can see how many likes it has.
  - I can see the post body.
  - I can see the username of each commentor.
  - I can see the comment each commentor left.

### CODE REVIEW

File: spec/features/user_index_spec.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  before do
+    visit users_path
+  end
+
+  it 'shows all users with their number of posts' do
+    users.each do |user|
+      expect(page).to have_content(user.name)
+      expect(page).to have_content("Number of Posts: #{user.posts.size}")
+    end
+  end
```
Comment (OyePriscilla) [inline spec/features/user_index_spec.rb]: - [ ] Please kindly add test that assert that profile picture is present as stated in the requirement. ℹ️ 

Untitled

File: spec/features/user_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  it 'shows the user profile with their posts' do
+    expect(page).to have_content(user.name)
+    expect(page).to have_content("Number of posts: #{user.posts.size}")
+    user.recent_posts.each do |post|
+      expect(page).to have_content(post.title)
+      expect(page).to have_content(post.text)
+      expect(page).to have_content("Comments: #{post.comments.size}")
+      expect(page).to have_content("Likes: #{post.likes_counter}")
+    end
+  end
```
Comment (OyePriscilla) [inline spec/features/user_show_spec.rb]: - [ ] Please kindly add the missing tests as highlighted in the screenshot below. 👇 
 
Untitled

File: spec/features/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+require 'rails_helper'
+
+RSpec.describe 'Post index page for a user', type: :feature do
+  let(:user) { create(:user) }
+  let!(:posts) { create_list(:post, 3, author: user) }
+  let!(:comments) { create_list(:comment, 3, post: posts.first) }
+
+  before do
+    visit user_posts_path(user)
+  end
```
Comment (OyePriscilla) [inline spec/features/post_index_spec.rb]: - [ ] Please kindly add the missing test for the post_index page as highlighted in the screenshot in accordance to the requirement. 🙏 

Untitled

File: spec/features/post_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+  before do
+    visit user_post_path(user, post)
+  end
+
+  it 'shows the post details with comments' do
+    expect(page).to have_content(post.title)
+    expect(page).to have_content(post.text)
+    expect(page).to have_content("Comments: #{post.comments.size}")
+    expect(page).to have_content("Likes: #{post.likes_counter}")
```
Comment (OyePriscilla) [inline spec/features/post_show_spec.rb]: - [ ] This test 👇 is not yet added to your implementation, kindly attend to it.  👍 

Untitled

**Review by OyePriscilla — CHANGES_REQUESTED** [review]
Hi @shm04 👏 👏 👏 👏, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

so-close

## Highlights
Write integration tests with Capybara gem. ✔️ 
Secure app from n+1 problems. ✔️ 

## Required Changes ♻️

**The link highlighted below from your README file was not accesible, kindly correct this for your repo to be more professional.**

Untitled

_Check the comments under the review also._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/user_show_spec.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  let(:user) { create(:user) }
+  let!(:posts) { create_list(:post, 5, author: user) }
+
+  before do
+    visit user_path(user)
+  end
+
+  it 'displays the user profile with their first 3 posts' do
+    expect(page).to have_content(user.name)
+    expect(page).to have_content("Number of posts: #{user.posts.size}")
```
Comment (cynthiainga) [inline spec/features/user_show_spec.rb:13]: - [ ] As mentioned by the previous reviewer you should add a test that we can see the user's profile picture 🙃 
- [ ] Also a test that we can see the user's first 3 posts, which means you should test the recent_posts method 🙃 🙃

File: spec/features/post_index_spec.rb
Comment (cynthiainga) [inline spec/features/post_index_spec.rb:1]: - [ ] As mentioned by the previous reviewer you should add a test for pagination, we can see a section for pagination if there are more posts than fit on the view 🙃 🙃

**Review by cynthiainga — CHANGES_REQUESTED** [review]
Hi @shm04, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.



 ## Highlights

- [ ] All implemented tests are passing ✔️ 
- [ ] No linters errors ✔️ 
- [ ] Correct Gitflow ✔️ 

 ## Required Changes ♻️

- [ ] Please kindly check the comments under the review 👇🏾 👇🏾 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were missed by the previous reviewer and addressing them will really improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by cynthiainga — APPROVED** [review]
Hi @shm04, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights

- [ ] All changes are made ✔️ 
- [ ] All tests are passing ✔️ 
- [ ] No linters errors ✔️ 
- [ ] Correct Gitflow ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were missed by the previous reviewer and addressing them will really improve your application._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**